planner/core/logical_plan_builder.go

```diff
@@ -3103,11 +3105,68 @@ func (b *PlanBuilder) buildWindowFunctionFrame(spec *ast.WindowSpec, orderByItem
 	return frame, err
 }
 
+func getAllByItems(spec *ast.WindowSpec) []*ast.ByItem {
+	var items []*ast.ByItem
+	if spec.PartitionBy != nil {
+		items = append(items, spec.PartitionBy.Items...)
+	}
+	if spec.OrderBy != nil {
+		items = append(items, spec.OrderBy.Items...)
+	}
+	return items
+}
+
+func restoreByItemText(item *ast.ByItem) string {
+	var sb strings.Builder
+	ctx := format.NewRestoreCtx(0, &sb)
+	err := item.Expr.Restore(ctx)
+	if err != nil {
+		return ""
+	}
+	return sb.String()
+}
+
+func compareItems(lItems []*ast.ByItem, rItems []*ast.ByItem) bool {
+	minLen := mathutil.Min(len(lItems), len(rItems))
+	for i := 0; i < minLen; i++ {
+		res := strings.Compare(restoreByItemText(lItems[i]), restoreByItemText(rItems[i]))
+		if res != 0 {
+			return res < 0
+		}
+		res = compareBool(lItems[i].Desc, rItems[i].Desc)
+		if res != 0 {
+			return res < 0
+		}
+	}
+	return len(lItems) < len(rItems)
+}
+
+type windowFuncs struct {
+	spec  *ast.WindowSpec
+	funcs []*ast.WindowFuncExpr
+}
+
+// sortWindowSpecs sorts the window specifications by reversed alphabetical order, then we could add less `Sort` operator
+// in physical plan because the window functions with the same partition by and order by clause will be at near places.
+func sortWindowSpecs(groupedFuncs map[*ast.WindowSpec][]*ast.WindowFuncExpr) []windowFuncs {
+	windows := make([]windowFuncs, 0, len(groupedFuncs))
+	for spec, funcs := range groupedFuncs {
+		windows = append(windows, windowFuncs{spec, funcs})
+	}
+	sort.SliceStable(windows, func(i, j int) bool {
+		lItems := getAllByItems(windows[i].spec)
```

Comment:
This function is very expensive, it allocates two `[]*ast.ByItem` slices in each compare function. Can we use a more efficient way?